ipa-core/src/net/client/mod.rs

```diff
@@ -39,13 +39,27 @@ pub enum ClientIdentity {
     /// Authenticate with an X.509 certificate or a certificate chain.
     ///
     /// This is only supported for HTTPS clients.
-    Certificate((Vec<Certificate>, PrivateKey)),
+    Certificate((Vec<CertificateDer<'static>>, PrivateKeyDer<'static>)),
 
     /// Do not authenticate nor claim a helper identity.
     #[default]
     None,
 }
 
+/// Rust-tls-types crate intentionally does not implement Clone on private key types in order to
```

Comment:
My two cents on this matter are:
a) `rustls::ClientConfig::builder with_client_auth_cert` requires to own its certificate.
b) We only clone the key 3 times. It's not like a certificate management system with different users or thousands of certificates in memory.

Reply:
> we have no control over how many copies of private key we hold in memory

Unless I'm misunderstanding something, I wouldn't put it this way. This makes it sound like a garbage-collected runtime making a copy of the key in memory each time garbage collection runs.

I do see how implementing `Clone` enables code using the API to do something inadvisable like make a new copy of the key on each request. But as I understand it, it's not license for the compiler to start spraying copies of the key everywhere in memory. We're using `Clone` to create three copies of the key for three HTTP clients which seems like a reasonable thing to do.

I see two changes we could make here (if we change anything):

1. Instead of implementing `Clone`, make a method like `ClientIdentity::clone_with_key` (like `PrivateKeyDer::clone_key`).
2. When we read the key (`fn server` in `src/bin/helper.rs`), create three duplicate identities, by cloning the `Vec<u8>` we read from the key file and calling `ClientIdentity::from_pks8` three times. There are also a few other code paths for creating identities that would need to change.

I'm fine with #&#x2060;1 although I don't think it's necessary. #&#x2060;2 seems silly but not actively harmful.

IMO fixing the typo in the name of the `from_pks8` method (should be `from_pkcs8`) is more important than addressing this.

Reply:
I did number 1 and and fixed the name to be `from_pkcs8`. 

For number 2 I'm not sure I understood. Couldn't it be to support the more general case when different helpers/clients can have different certificates and such? In that CLI it seems redundant, but on a more general use case, it might make sense to keep the code as is?

Reply:
> I do see how implementing Clone enables code using the API to do something inadvisable like make a new copy of the key on each request. 

+1

>it's not license for the compiler to start spraying copies of the key everywhere in memory

Yea, my main concern is with application code, I don't worry too much about compiler going wild.

> We're using Clone to create three copies of the key for three HTTP clients which seems like a reasonable thing to do.

As long as we trust ourselves to do the right thing. I do see `derive(Clone)` as a footgun for inadvertently replicating private key bytes. 


I don't think this issue requires immediate attention though

Reply:
> Couldn't it be to support the more general case when different helpers/clients can have different certificates and such? In that CLI it seems redundant, but on a more general use case, it might make sense to keep the code as is?

I was thinking of changing `MpcHelperClient::from_conf` (which already returns three clients) to take an argument of type `[ClientIdentity; 3]` instead of `ClientIdentity`. Then it wouldn't need to clone client identities internally. As a side effect of making that change, it would also be more flexible (support using a different identity for each of the clients), but I don't foresee a scenario where we care about having that flexibility.